src/mqtt.rs

```diff
@@ -231,19 +231,27 @@ pub enum ChannelData {
     Shutdown,
 }
 
+pub type Receiver = broadcast::Receiver<ChannelData>;
 pub type Sender = broadcast::Sender<ChannelData>;
 
 pub struct Mqtt {
     config: ConfigWrapper,
     shutdown: bool,
     channels: Channels,
+    receiver: std::cell::Cell<Option<Receiver>>,
 }
 
 impl Mqtt {
     pub fn new(config: ConfigWrapper, channels: Channels) -> Self {
+        // Create the receiver at new() time, rather than start(), so things
+        // wanting to send to MQTT queue messages, rather than failing with an
```

Comment:
I was surprised to learn that these tokio broadcast channels have a fixed-size buffer (set to 512 in src/channels.rs) and that messages are dropped when the buffer is full, rather than send() async-blocking until a recv() call happened on the other side.

This design works with either approach, but where it matters most is when MQTT is disabled, where if it were the latter case, we'd have to do extra work to avoid blocking forever.

As-is, rather than the message being discarded with an error (there are examples in the codebase of that error being picked up on), the send() will now succeed. 512 messages will be kept in RAM.

It would be better for the sender to notice that MQTT is disabled and not try to send the message to begin with. I can add that to this PR if desired.

I guess a similar situation prevails with influx?

Reply:
512 was a somewhat arbitrary choice that I just hoped we'd never run into :) Its enough for all the input registers to be queued and then we pretty much just hope they get cleared before more come in. In normal circumstances they're gone in a second or so.

I thought most places that tried to send to MQTT already checked via something like `self.config.mqtt().enabled()` - if any don't then that should probably be added.

This setup all starts to get a bit messy when things aren't behaving, I'm not sure what the behaviour might be if we async-block on sending to a channel anywhere? If the coordinator is blocked then we'll probably run into all sorts of other issues. Maybe we're reaching the limitations of this design. I'm pretty open to suggestions on improvements here, the use of channels in this way was a very early decision while learning rudimentary Rust :)

Also looks like the `broadcast` channels currently in use don't support waiting to send, we'd have to swap to `mpsc` to get that: https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.send - but I don't think that would be a problem. This one also supports unbounded channels so we could keep queueing forever but that feels like a memory balloon waiting to happen. Would be better to deterministically decide what happens if things aren't working and messages aren't being consumed.

Reply:
I was just chewing the fat here TBH. My own rust is pretty early-stage, but the general idea seems to be to design these things out through the type system, as with the idea about removing `Option` above.

Do you want us to make further progress on this before merging this PR, or are you OK with it going in as-is?

If we need further changes, one thought I had was a null object sort of thing - basically a discard sink that could be the subscriber to the `to_mqtt` channel when mqtt is disabled. Something similar for influx and the other disable-able components too. That would allow the `if ... enabled()` conditionals to go away, along with the `bail!`s around the closed channel messages that are currently here. That could go into this PR if you don't like the buffers filling up when mqtt is disabled? Just let me know.

Silently dropping messages when the buffer is full is :grimacing:,  and switching to unbuffered, or at least blocks-when-full, channels fixes that, but it has all sorts of implications for the design of the rest of the system to keep it from deadlocking anything :sweat_smile:. It's a thought for how to go in the future, but I'd hate for the holding register publish stuff to be be (ahahaha) blocked behind

Reply:
Ah, the `bail` I had in mind is in `src/coordinator/commands/time_register_ops.rs`, so would only ever be called if mqtt were enabled anyway (although that's not guaranteed by construction, it happens to be true).

The only way it (and the others in `mod.rs`) could fire now with mqtt enabled is during shutdown, I think - this PR designs out the possibility that they could fire while mqtt is enabled and `stop` hasn't been called. Could probably design that out too, as part of making the `stop` methods go away.